deepmd/dpmodel/utils/network.py

```diff
@@ -957,3 +957,167 @@ def deserialize(cls, data: dict) -> "NetworkCollection":
         check_version_compatibility(data.pop("@version", 1), 1, 1)
         data.pop("@class", None)
         return cls(**data)
+
+
+def aggregate(
+    data: np.ndarray,
+    owners: np.ndarray,
+    average=True,
+    num_owner=None,
+):
+    """
+    Aggregate rows in data by specifying the owners.
+
+    Parameters
+    ----------
+    data : data tensor to aggregate [n_row, feature_dim]
+    owners : specify the owner of each row [n_row, 1]
+    average : if True, average the rows, if False, sum the rows.
+        Default = True
+    num_owner : the number of owners, this is needed if the
+        max idx of owner is not presented in owners tensor
+        Default = None
+
+    Returns
+    -------
+    output: [num_owner, feature_dim]
+    """
+    xp = array_api_compat.array_namespace(data, owners)
+
+    def add_at(x, indices, values):
+        unique_ids = xp.unique(indices)
+        for i in unique_ids:
+            mask = xp.where(indices == i, 1, 0)
+            mask = xp.expand_dims(mask, axis=1) if len(values.shape) != 1 else mask
+            selected = values * mask
+            summed = xp.sum(selected, axis=0)
+            x[i] = x[i] + summed
+
+        return x
+
+    def bincount(x, weights=None, minlength=0):
+        if weights is None:
+            weights = xp.ones_like(x)
+        result = xp.zeros((max(minlength, int(xp.max(x)) + 1),), dtype=weights.dtype)
+        result = add_at(result, x, weights)
+        return result
+
+    if hasattr(xp, "bincount"):
```

Comment:
```suggestion
    if array_api_compat.is_numpy_namespace(xp) or array_api_compat.is_jax_namespace(xp):
```

I suggest checking it explicitly... We cannot assume a new array library to have the same features as NumPy, even if its method name is the same.